<a href="https://colab.research.google.com/github/oonid/Coursera_Capstone/blob/main/Segmenting_Clustering_Neighbourhoods_Toronto.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Applied Data Science Capstone: Segmenting and Clustering Neighbourhoods in Toronto

Applied Data Science Capstone, week 3. Coursera. IBM.


In [1]:
# libraries

import pandas as pd
import numpy as np
import requests
import io

# from pandas.io.json import json_normalize  # this is deprecated, use below
from pandas import json_normalize  # tranform JSON file into a pandas dataframe

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

# import k-means from clustering stage
from sklearn.cluster import KMeans

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# Foursquare CLIENT_ID and CLIENT_SECRET will be saved as pass
from getpass import getpass

print('Libraries imported.')

Libraries imported.


# Assignment 1

*   Scaping data from [list of postal codes of Canada](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M).
*   Using `pandas` library to read the HTML table and convert it into DataFrame.
*   `pandas` function `read_html` detect 3 tables on the page, so using **assumption** that the postal code is the first table (index 0).
*   Change the column name 'Postal Code' to 'PostalCode' to follow the guidance.
*   Remove rows that has **Not assigned** values of Borough and reset the index. As the effect, multiple rows that have the same Postal Code will be combined with comma as separator, for example Postal Code **M5A**.

In [2]:
tables = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
print('total found tables in page: {}'.format(len(tables)))

postal_code = tables[0]  # assumption the postal code table is the first table
print('set the first found table as postal_code dataframe, initial shape: {}'
    .format(postal_code.shape))

postal_code.rename(columns={'Postal Code': 'PostalCode'}, inplace=True)

not_assigned_borough = postal_code[postal_code['Borough'] == 'Not assigned']
print('remove rows with Not assigned Borough ({} rows).'
    .format(not_assigned_borough.shape[0]))

borough_not_assigned_neighbourhood = \
    postal_code[(postal_code['Borough'] != 'Not assigned') & \
                (postal_code['Neighbourhood'] == 'Not assigned')]
print('set Not assigned Neighbourhood with Borough values ({} rows).'
    .format(borough_not_assigned_neighbourhood.shape[0]))

# ignore data with "Not assigned" Borough
postal_code = \
    postal_code[postal_code['Borough'] != 'Not assigned'].reset_index(drop=True)

# make sure the postal code M5A combined
postal_code_m5a = postal_code[postal_code['PostalCode'] == 'M5A']
print('total rows for M5A: {} row(s). 1 row means already combined.'
    .format(postal_code_m5a.shape[0]))

# output the postal_code dataframe
print('Postal Code dataframe:')
postal_code

total found tables in page: 3
set the first found table as postal_code dataframe, initial shape: (180, 3)
remove rows with Not assigned Borough (77 rows).
set Not assigned Neighbourhood with Borough values (0 rows).
total rows for M5A: 1 row(s). 1 row means already combined.
Postal Code dataframe:


PostalCode  ...                                      Neighbourhood
0          M3A  ...                                          Parkwoods
1          M4A  ...                                   Victoria Village
2          M5A  ...                          Regent Park, Harbourfront
3          M6A  ...                   Lawrence Manor, Lawrence Heights
4          M7A  ...        Queen's Park, Ontario Provincial Government
..         ...  ...                                                ...
98         M8X  ...      The Kingsway, Montgomery Road, Old Mill North
99         M4Y  ...                               Church and Wellesley
100        M7Y  ...  Business reply mail Processing Centre, South C...
101        M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
102        M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

In [3]:
# output the final data frame shape for Assignment 1
postal_code.shape

(103, 3)

# Assignment 2

After build DataFrame of te postal code of each neighbourhood along with the boroughname and neighbourhood name, in order to utilize the Foursquare location data, get the latitude and longitude coordinates of each neighbourhood.

#### 1st Try Using geocoder, the Result is Failed

The application keep looping without any result, so we stop using this library. FAILED.

*   install geocoder package (free, but not reliable)
```
!pip install geocoder
```

*   import geocoder library after successfully installed
```
import geocoder
```

*   get latitude and longitude from geocoder library
```
# iterate for every PostalCode and try to get the coordinate via geocoder
for postal_code_value in postal_code['PostalCode']:
    print('process data for PostalCode: {}'.format(postal_code_value))

    # initialize your variable to None
    lat_lng_coords = None

    # loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.google('{}, Toronto, Ontario'.format(postal_code_value))
        lat_lng_coords = g.latlng
        print('.', end='', flush=True)
 
    latitude = lat_lng_coords[0]
    longitude = lat_lng_coords[1]

    print('PostalCode {} latitude: {} longitude: {}'
        .format(postal_code_value, latitude, longitude))
```


#### As the alternative of geocoder, get the latitude and longitude from [predefined geospatial data](http://cocl.us/Geospatial_data).

In [4]:
# the URL contains redirection (HTTP 308) so get the string content first
s = requests.get('http://cocl.us/Geospatial_data').content
geo_data = pd.read_csv(io.StringIO(s.decode('latin-1')))
geo_data.rename(columns={'Postal Code': 'PostalCode'}, inplace=True)
geo_data

PostalCode   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
..         ...        ...        ...
98         M9N  43.706876 -79.518188
99         M9P  43.696319 -79.532242
100        M9R  43.688905 -79.554724
101        M9V  43.739416 -79.588437
102        M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [5]:
# join geo_data to postal_code to append Latitude and Longitude columns
neighbourhoods = pd.merge(postal_code, geo_data, on=['PostalCode'])
neighbourhoods

PostalCode           Borough  ...   Latitude  Longitude
0          M3A        North York  ...  43.753259 -79.329656
1          M4A        North York  ...  43.725882 -79.315572
2          M5A  Downtown Toronto  ...  43.654260 -79.360636
3          M6A        North York  ...  43.718518 -79.464763
4          M7A  Downtown Toronto  ...  43.662301 -79.389494
..         ...               ...  ...        ...        ...
98         M8X         Etobicoke  ...  43.653654 -79.506944
99         M4Y  Downtown Toronto  ...  43.665860 -79.383160
100        M7Y      East Toronto  ...  43.662744 -79.321558
101        M8Y         Etobicoke  ...  43.636258 -79.498509
102        M8Z         Etobicoke  ...  43.628841 -79.520999

[103 rows x 5 columns]

In [6]:
neighbourhoods.shape

(103, 5)

# Assignment 3

Explore and cluster the neighborhoods in Toronto. Work with only boroughs that **contains the word Toronto** and then replicate the same analysis we did to the New York City data.


The Canada city dataset has all 10 boroughs and 103 neighbourhoods.

In [7]:
print('The dataframe has {} boroughs and {} neighbourhoods.'.format(
        len(neighbourhoods['Borough'].unique()),
        neighbourhoods.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighbourhoods.


Work with only boroughs that **contains the word Toronto**.

In [8]:
toronto_data = \
    neighbourhoods[neighbourhoods['Borough'].str.contains('Toronto')] \
    .reset_index(drop=True)
toronto_data

PostalCode           Borough  ...   Latitude  Longitude
0         M5A  Downtown Toronto  ...  43.654260 -79.360636
1         M7A  Downtown Toronto  ...  43.662301 -79.389494
2         M5B  Downtown Toronto  ...  43.657162 -79.378937
3         M5C  Downtown Toronto  ...  43.651494 -79.375418
4         M4E      East Toronto  ...  43.676357 -79.293031
5         M5E  Downtown Toronto  ...  43.644771 -79.373306
6         M5G  Downtown Toronto  ...  43.657952 -79.387383
7         M6G  Downtown Toronto  ...  43.669542 -79.422564
8         M5H  Downtown Toronto  ...  43.650571 -79.384568
9         M6H      West Toronto  ...  43.669005 -79.442259
10        M5J  Downtown Toronto  ...  43.640816 -79.381752
11        M6J      West Toronto  ...  43.647927 -79.419750
12        M4K      East Toronto  ...  43.679557 -79.352188
13        M5K  Downtown Toronto  ...  43.647177 -79.381576
14        M6K      West Toronto  ...  43.636847 -79.428191
15        M4L      East Toronto  ...  43.668999 -79.315572
16        M5L  Downtown Toronto  ...  43.648198 -79.379817
17        M4M      East Toronto  ...  43.659526 -79.340923
18        M4N   Central Toronto  ...  43.728020 -79.388790
19        M5N   Central Toronto  ...  43.711695 -79.416936
20        M4P   Central Toronto  ...  43.712751 -79.390197
21        M5P   Central Toronto  ...  43.696948 -79.411307
22        M6P      West Toronto  ...  43.661608 -79.464763
23        M4R   Central Toronto  ...  43.715383 -79.405678
24        M5R   Central Toronto  ...  43.672710 -79.405678
25        M6R      West Toronto  ...  43.648960 -79.456325
26        M4S   Central Toronto  ...  43.704324 -79.388790
27        M5S  Downtown Toronto  ...  43.662696 -79.400049
28        M6S      West Toronto  ...  43.651571 -79.484450
29        M4T   Central Toronto  ...  43.689574 -79.383160
30        M5T  Downtown Toronto  ...  43.653206 -79.400049
31        M4V   Central Toronto  ...  43.686412 -79.400049
32        M5V  Downtown Toronto  ...  43.628947 -79.394420
33        M4W  Downtown Toronto  ...  43.679563 -79.377529
34        M5W  Downtown Toronto  ...  43.646435 -79.374846
35        M4X  Downtown Toronto  ...  43.667967 -79.367675
36        M5X  Downtown Toronto  ...  43.648429 -79.382280
37        M4Y  Downtown Toronto  ...  43.665860 -79.383160
38        M7Y      East Toronto  ...  43.662744 -79.321558

[39 rows x 5 columns]

#### Use geopy library to get the latitude and longitude values of Toronto City.


In [9]:
address = 'Toronto City, Canada'

geolocator = Nominatim(user_agent="trt_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'
    .format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


#### Create a map of New York with neighborhoods superimposed on top.


In [10]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in \
        zip(toronto_data['Latitude'], toronto_data['Longitude'],
            toronto_data['Borough'], toronto_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## Complete the Toronto dataset with information from Foursquare

#### Define Foursquare Credentials and Version


In [11]:
CLIENT_ID = getpass('Foursquare ID: ') # your Foursquare ID
CLIENT_SECRET = getpass('Foursquare Secret: ') # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
# print('CLIENT_SECRET:' + CLIENT_SECRET)

Foursquare ID: ··········
Foursquare Secret: ··········
Your credentails:
CLIENT_ID: I2ZHJ20LEPMTBLGLXWJVRBAG00MW3BQFDK2OA3QYMEFX41YG


#### Let's explore the first neighborhood in our dataframe.


Get the neighborhood's name.

In [12]:
toronto_data.loc[0, 'Neighbourhood']

'Regent Park, Harbourfront'

Get the neighborhood's latitude and longitude values.

In [14]:
neighbourhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighbourhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighbourhood_name = toronto_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'
    .format(neighbourhood_name, neighbourhood_latitude, neighbourhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


#### Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.

In [15]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)

# url # display URL

Send the GET request and examine the resutls

In [16]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '6043470b7890fd59af4e8762'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-54ea41ad498e9a11e9e13308-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/bakery_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d16a941735',
         'name': 'Bakery',
         'pluralName': 'Bakeries',
         'primary': True,
         'shortName': 'Bakery'}],
       'id': '54ea41ad498e9a11e9e13308',
       'location': {'address': '362 King St E',
        'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'crossStreet': 'Trinity St',
        'distance': 143,
        'formattedAddress': ['362 King St E (Trinity St)',
         'Toronto ON M5A 1K9',
         'Canada'],
        'labeledLatLngs': [{'label': 'display',
 

Borrow the **get_category_type** function from the Foursquare lab, all the information is in the _items_ key.

In [17]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a _pandas_ dataframe.


In [18]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat',
                    'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = \
    nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

nearby_venues


47 venues were returned by Foursquare.


name  ...        lng
0                       Roselle Desserts  ... -79.362017
1                          Tandem Coffee  ... -79.361809
2                 Cooper Koo Family YMCA  ... -79.358008
3                    Body Blitz Spa East  ... -79.359874
4                         Impact Kitchen  ... -79.356980
5               Dominion Pub and Kitchen  ... -79.358967
6                        Corktown Common  ... -79.356211
7                     Morning Glory Cafe  ... -79.361149
8                     The Extension Room  ... -79.359725
9       The Distillery Historic District  ... -79.359323
10                   SOMA chocolatemaker  ... -79.358127
11              Distillery Sunday Market  ... -79.361832
12                Figs Breakfast & Lunch  ... -79.364503
13                   Brick Street Bakery  ... -79.359539
14                                  Arvo  ... -79.361442
15                       Sumach Espresso  ... -79.359515
16                        Rooster Coffee  ... -79.365609
17                        Underpass Park  ... -79.354806
18               Dark Horse Espresso Bar  ... -79.357078
19  Young Centre for the Performing Arts  ... -79.357593
20                             Starbucks  ... -79.364917
21                      Souvlaki Express  ... -79.364438
22                Parliament Square Park  ... -79.362195
23            Cluny Bistro & Boulangerie  ... -79.357843
24                       ODIN Cafe + Bar  ... -79.356503
25                              Cacao 70  ... -79.360723
26                       The Yoga Lounge  ... -79.364955
27                           Caffe Furbo  ... -79.358849
28                       Alumnae Theatre  ... -79.364753
29                    Soulpepper Theatre  ... -79.357615
30                       Berkeley Church  ... -79.365873
31                    John Fluevog Shoes  ... -79.359436
32                                 Izumi  ... -79.360153
33           The Sweet Escape Patisserie  ... -79.358709
34     Ontario Spring Water Sake Company  ... -79.360073
35                            The Aviary  ... -79.354662
36                          Arta Gallery  ... -79.361222
37                                Vistek  ... -79.359667
38           The Abnormal Beauty Company  ... -79.361005
39                       TD Canada Trust  ... -79.356336
40                       Savoury Grounds  ... -79.358970
41                        The Beer Store  ... -79.357463
42         Residence & Conference Centre  ... -79.357040
43                      The Healthy Road  ... -79.357119
44                             Wine Rack  ... -79.356928
45                                 FUEL+  ... -79.355867
46                            GW General  ... -79.357538

[47 rows x 4 columns]

## 2. Explore Neighborhoods in Toronto

#### Create a function to repeat the same process to all the neighborhoods in Toronto


In [19]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}' \
              '&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame(
        [item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called _toronto_venues_.

In [20]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighbourhood'],
                                 latitudes=toronto_data['Latitude'],
                                 longitudes=toronto_data['Longitude'])


Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

#### Let's check the size of the resulting dataframe


In [21]:
print(toronto_venues.shape)
toronto_venues

(1602, 7)


Neighbourhood  ...        Venue Category
0                             Regent Park, Harbourfront  ...                Bakery
1                             Regent Park, Harbourfront  ...           Coffee Shop
2                             Regent Park, Harbourfront  ...   Distribution Center
3                             Regent Park, Harbourfront  ...                   Spa
4                             Regent Park, Harbourfront  ...            Restaurant
...                                                 ...  ...                   ...
1597  Business reply mail Processing Centre, South C...  ...           Yoga Studio
1598  Business reply mail Processing Centre, South C...  ...               Butcher
1599  Business reply mail Processing Centre, South C...  ...            Smoke Shop
1600  Business reply mail Processing Centre, South C...  ...  Gym / Fitness Center
1601  Business reply mail Processing Centre, South C...  ...      Recording Studio

[1602 rows x 7 columns]

Let's check how many venues were returned for each neighborhood


In [22]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  ...  Venue Category
Neighbourhood                                                               ...                
Berczy Park                                                             58  ...              58
Brockton, Parkdale Village, Exhibition Place                            24  ...              24
Business reply mail Processing Centre, South Ce...                      17  ...              17
CN Tower, King and Spadina, Railway Lands, Harb...                      15  ...              15
Central Bay Street                                                      59  ...              59
Christie                                                                15  ...              15
Church and Wellesley                                                    75  ...              75
Commerce Court, Victoria Hotel                                         100  ...             100
Davisville                                                              32  ...              32
Davisville North                                                         8  ...               8
Dufferin, Dovercourt Village                                            15  ...              15
First Canadian Place, Underground city                                 100  ...             100
Forest Hill North & West, Forest Hill Road Park                          4  ...               4
Garden District, Ryerson                                               100  ...             100
Harbourfront East, Union Station, Toronto Islands                      100  ...             100
High Park, The Junction South                                           25  ...              25
India Bazaar, The Beaches West                                          21  ...              21
Kensington Market, Chinatown, Grange Park                               63  ...              63
Lawrence Park                                                            3  ...               3
Little Portugal, Trinity                                                42  ...              42
Moore Park, Summerhill East                                              3  ...               3
North Toronto West, Lawrence Park                                       19  ...              19
Parkdale, Roncesvalles                                                  14  ...              14
Queen's Park, Ontario Provincial Government                             30  ...              30
Regent Park, Harbourfront                                               47  ...              47
Richmond, Adelaide, King                                                95  ...              95
Rosedale                                                                 4  ...               4
Roselawn                                                                 3  ...               3
Runnymede, Swansea                                                      34  ...              34
St. James Town                                                          80  ...              80
St. James Town, Cabbagetown                                             47  ...              47
Stn A PO Boxes                                                          99  ...              99
Studio District                                                         36  ...              36
Summerhill West, Rathnelly, South Hill, Forest ...                      14  ...              14
The Annex, North Midtown, Yorkville                                     21  ...              21
The Beaches                                                              5  ...               5
The Danforth West, Riverdale                                            43  ...              43
Toronto Dominion Centre, Design Exchange                               100  ...             100
University of Toronto, Harbord                                          32  ...              32

[39 rows x 6 columns]

#### Let's find out how many unique categories can be curated from all the returned venues

In [23]:
print('There are {} uniques categories.'
    .format(len(toronto_venues['Venue Category'].unique())))

There are 230 uniques categories.


## 3. Analyze Each Neighborhood


In [24]:
# one hot encoding
toronto_onehot = \
    pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighbourhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot

Neighbourhood  ...  Yoga Studio
0                             Regent Park, Harbourfront  ...            0
1                             Regent Park, Harbourfront  ...            0
2                             Regent Park, Harbourfront  ...            0
3                             Regent Park, Harbourfront  ...            0
4                             Regent Park, Harbourfront  ...            0
...                                                 ...  ...          ...
1597  Business reply mail Processing Centre, South C...  ...            1
1598  Business reply mail Processing Centre, South C...  ...            0
1599  Business reply mail Processing Centre, South C...  ...            0
1600  Business reply mail Processing Centre, South C...  ...            0
1601  Business reply mail Processing Centre, South C...  ...            0

[1602 rows x 231 columns]

In [25]:
toronto_onehot.shape

(1602, 231)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [26]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  ...  Yoga Studio
0                                         Berczy Park  ...     0.000000
1        Brockton, Parkdale Village, Exhibition Place  ...     0.041667
2   Business reply mail Processing Centre, South C...  ...     0.058824
3   CN Tower, King and Spadina, Railway Lands, Har...  ...     0.000000
4                                  Central Bay Street  ...     0.016949
5                                            Christie  ...     0.000000
6                                Church and Wellesley  ...     0.026667
7                      Commerce Court, Victoria Hotel  ...     0.000000
8                                          Davisville  ...     0.000000
9                                    Davisville North  ...     0.000000
10                       Dufferin, Dovercourt Village  ...     0.000000
11             First Canadian Place, Underground city  ...     0.000000
12    Forest Hill North & West, Forest Hill Road Park  ...     0.000000
13                           Garden District, Ryerson  ...     0.000000
14  Harbourfront East, Union Station, Toronto Islands  ...     0.000000
15                      High Park, The Junction South  ...     0.000000
16                     India Bazaar, The Beaches West  ...     0.000000
17          Kensington Market, Chinatown, Grange Park  ...     0.000000
18                                      Lawrence Park  ...     0.000000
19                           Little Portugal, Trinity  ...     0.023810
20                        Moore Park, Summerhill East  ...     0.000000
21                  North Toronto West, Lawrence Park  ...     0.052632
22                             Parkdale, Roncesvalles  ...     0.000000
23        Queen's Park, Ontario Provincial Government  ...     0.033333
24                          Regent Park, Harbourfront  ...     0.021277
25                           Richmond, Adelaide, King  ...     0.000000
26                                           Rosedale  ...     0.000000
27                                           Roselawn  ...     0.000000
28                                 Runnymede, Swansea  ...     0.029412
29                                     St. James Town  ...     0.000000
30                        St. James Town, Cabbagetown  ...     0.000000
31                                     Stn A PO Boxes  ...     0.010101
32                                    Studio District  ...     0.027778
33  Summerhill West, Rathnelly, South Hill, Forest...  ...     0.000000
34                The Annex, North Midtown, Yorkville  ...     0.000000
35                                        The Beaches  ...     0.000000
36                       The Danforth West, Riverdale  ...     0.023256
37           Toronto Dominion Centre, Design Exchange  ...     0.000000
38                     University of Toronto, Harbord  ...     0.031250

[39 rows x 231 columns]

In [27]:
toronto_grouped.shape

(39, 231)

#### Let's print each neighborhood along with the top 5 most common venues


In [28]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0     Coffee Shop  0.09
1    Cocktail Bar  0.05
2          Bakery  0.05
3  Farmers Market  0.03
4        Pharmacy  0.03


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.12
1  Breakfast Spot  0.08
2     Coffee Shop  0.08
3       Nightclub  0.08
4     Yoga Studio  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
           venue  freq
0    Yoga Studio  0.06
1  Auto Workshop  0.06
2        Brewery  0.06
3        Butcher  0.06
4     Skate Park  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.13
1   Airport Service  0.13
2  Airport Terminal  0.13
3           Airport  0.07
4               Bar  0.07


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.19
1  Italian Restaurant  0.05
2                C

#### Let's put that into a _pandas_ dataframe


First, let's write a function to sort the venues in descending order.


In [29]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.


In [30]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = \
        return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood  ... 10th Most Common Venue
0                                        Berczy Park  ...                   Café
1       Brockton, Parkdale Village, Exhibition Place  ...      Convenience Store
2  Business reply mail Processing Centre, South C...  ...                 Garden
3  CN Tower, King and Spadina, Railway Lands, Har...  ...               Boutique
4                                 Central Bay Street  ...  Portuguese Restaurant

[5 rows x 11 columns]

## 4. Cluster Neighborhoods


Run _k_-means to cluster the neighborhood into 5 clusters.


In [31]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0) \
    .fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [32]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = \
    toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'),
                        on='Neighbourhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough  ... 9th Most Common Venue  10th Most Common Venue
0        M5A  Downtown Toronto  ...      Greek Restaurant               Wine Shop
1        M7A  Downtown Toronto  ...         Smoothie Shop     Japanese Restaurant
2        M5B  Downtown Toronto  ...                 Hotel    Fast Food Restaurant
3        M5C  Downtown Toronto  ...            Restaurant      Italian Restaurant
4        M4E      East Toronto  ...               Dog Run        Doner Restaurant

[5 rows x 16 columns]

Finally, let's visualize the resulting clusters


In [33]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in \
        zip(toronto_merged['Latitude'], toronto_merged['Longitude'],
            toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters


Examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, then assign a name to each cluster.


#### Cluster 1: Shopping and Restaurant


In [34]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0,
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ...  10th Most Common Venue
0   Downtown Toronto  ...               Wine Shop
1   Downtown Toronto  ...     Japanese Restaurant
2   Downtown Toronto  ...    Fast Food Restaurant
3   Downtown Toronto  ...      Italian Restaurant
4       East Toronto  ...        Doner Restaurant
5   Downtown Toronto  ...                    Café
6   Downtown Toronto  ...   Portuguese Restaurant
7   Downtown Toronto  ...                 Dog Run
8   Downtown Toronto  ...      Seafood Restaurant
9       West Toronto  ...                    Pool
10  Downtown Toronto  ...             Music Venue
11      West Toronto  ...               Juice Bar
12      East Toronto  ...     Japanese Restaurant
13  Downtown Toronto  ...      Italian Restaurant
14      West Toronto  ...       Convenience Store
15      East Toronto  ...              Restaurant
16  Downtown Toronto  ...            Cocktail Bar
17      East Toronto  ...            Neighborhood
20   Central Toronto  ...               Gastropub
22      West Toronto  ...      Italian Restaurant
23   Central Toronto  ...              Restaurant
24   Central Toronto  ...               BBQ Joint
25      West Toronto  ...               Bookstore
26   Central Toronto  ...          Farmers Market
27  Downtown Toronto  ...                Beer Bar
28      West Toronto  ...     Indie Movie Theater
30  Downtown Toronto  ...                    Park
31   Central Toronto  ...                    Bank
32  Downtown Toronto  ...                Boutique
34  Downtown Toronto  ...          Clothing Store
35  Downtown Toronto  ...                  Office
36  Downtown Toronto  ...           Deli / Bodega
37  Downtown Toronto  ...                     Pub
38      East Toronto  ...                  Garden

[34 rows x 12 columns]

#### Cluster 2


In [35]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1,
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ...  10th Most Common Venue
18  Central Toronto  ...       Electronics Store

[1 rows x 12 columns]

#### Cluster 3: Garden and Services


In [36]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2,
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ...       10th Most Common Venue
19  Central Toronto  ...  Eastern European Restaurant

[1 rows x 12 columns]

#### Cluster 4: Jewelry Store


In [37]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3,
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ...  10th Most Common Venue
21  Central Toronto  ...             Escape Room

[1 rows x 12 columns]

#### Cluster 5: Park and Trail


In [38]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4,
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ...       10th Most Common Venue
29   Central Toronto  ...          Dumpling Restaurant
33  Downtown Toronto  ...  Eastern European Restaurant

[2 rows x 12 columns]